kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,486 @@
+                # used purely for the unit tests to validate which packages
+                # were deleted.
+                all_packages_deleted.append(package)
+
+                # Track that this package was deleted -- used in the parent for
+                # loop to give the user a final tally of the packages that
+                # were kept, and that were deleted.
+                packages_deleted.append(package)
+
+            # Print out the packages that were not deleted and left in the repo
+            all_files = ['%s/%s' %
+                         (package['distro_version'], package['filename'])
+                         for package in package_versions]
+            deleted_files = ['%s/%s' %
+                             (package['distro_version'], package['filename'])
+                             for package in packages_deleted]
+            files_left = list(set(all_files) - set(deleted_files))
+            self.log.debug('%s remaining packages: %s' % (name, files_left))
+
+        raise gen.Return(all_packages_deleted)
+
+
+class Delete(PackagecloudBase):
+
+    """Deletes packages from a PackageCloud repo.
+
+    Searches for packages that match the `packages_to_delete` regex pattern and
+    deletes them.  If `number_to_keep` is set, we always at least this number
+    of versions of the given package intact in the repo.  Also if
+    `number_to_keep` is set, the older versions of a package (based on upload
+    time) packages will be deleted first effectively leaving newer packages
+    in the repo.
+
+    **Options**
+
+    :number_to_keep:
+      Keep at least this number of each package
+      (defaults to *0*)
+
+    :packages_to_delete:
+      Regex of packages to delete, e.g. pkg1|pkg2
+
+    :repo:
+      Which packagecloud repo to delete from
+
+    **Examples**
+
+    .. code-block:: json
+
+      { "desc": "packagecloud Delete example",
+        "actor": "packagecloud.Delete",
+        "options": {
+          "number_to_keep": 10,
+          "packages_to_delete": "deleteme",
+          "repo": "test"
+        }
+      }
+
+    """
+
+    all_options = {
+        'number_to_keep': (
+            int, 0,
+            'Keep at least this number of each package'),
+        'packages_to_delete': (
+            str, REQUIRED,
+            'Regex of packages to delete, e.g. pkg1|pkg2'),
+        'repo': (
+            str, REQUIRED,
+            'Which packagecloud repo to delete from'),
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Check required environment variables."""
+        super(Delete, self).__init__(*args, **kwargs)
+
+        try:
+            re.compile(self.option('packages_to_delete'))
+        except re.error:
+            raise exceptions.InvalidOptions(
+                'packages_to_delete is an invalid regex')
+
+    @gen.coroutine
+    def _execute(self):
+        """Deletes all packages that match the `packages_to_delete` pattern"""
+        yield self._delete(
```

Comment:
Given the complexity of the loop, I am OK with it being the way it is now. If we want to do what you suggest, it means that we have to create several more methods and handle passing a ton of data around.